### PR TITLE
ENH: Added kstwo probability distribution for two-sided one-sample Kolmogorov-Smirnov tests.

### PR DESCRIPTION
**Pull request by pvanmulbregt**
[Replaces gh-9716 with a rebase.]
Add the `kstwo` probability distribution, for use in Kolmogorov-Smirnov tests.
Already existing are `ksone` and `kstgwobign`.
`ksone.cdf(n, x)` returns the one-sided Kolmogorov-Smirnov `Pr(Dn+ <= x)`.
`kstwobign.cdf(z)` returns the asymptotic two-sided Kolmogorov-Smirnov `Pr(sqrt(n)*Dn <= z)`, suitable for large values of `n`.
The new `kstwo.cdf(n, x`) returns the exact two-sided Kolmogorov-Smirnov `Pr(Dn <= x)`.

The exact computation of the two-sided `Pr(Dn <= x)` can be expensive. A number of exact and approximate algorithms were developed in the 1960s and 70s. Work of Simard and L'Ecuyer [1] provided a recipe for when to use each one.

This PR adds the `kstwo` distribution. Two more PRs are planned: one to compute exact probabilities for the one- and two-sided two-sample test statistics in `stats.ks_2samp`; the other to address usage of `ksone`/`kstwobign` in `stats.kstest`.

[1] Simard, R., L'Ecuyer, P. (2011)
"Computing the Two-Sided Kolmogorov-Smirnov Distribution",
Journal of Statistical Software, Vol 39, 11, 1-18.

### CODE REVIEW

**Comment by chrisb83** [comment]
Looks like a very nice addition to `stats`. Are there differences to the code linked in the paper (http://www.iro.umontreal. ca/~simardr/ksdir/)?

**Comment by pvanmulbregt** [comment]
I worked from the paper[1], avoiding the java/C code in http://www.iro.umontreal.ca/~simardr/ksdir/.  The implementations in this gh-9719  started from the basic algorithm as described in [1], but then evolved using numpy/scipy library calls and other optimizations.  Hence I would expect the code in the referenced java/C code to be similar in spirit to this PR, but perhaps a little different.

**Comment by pvanmulbregt** [comment]
The one test failure is unrelated to this PR: the scipy.scipy "Windows Python36-32bit-full" build (Issue gh-9724, addressed  by gh-9725).

**Comment by pvanmulbregt** [comment]
The 2 test failures in scipy.scipy are timeouts involving kstwo, which need understanding and addressing.

**Comment by WarrenWeckesser** [comment]
@pvanmulbregt, I'm currently trying to get @chrisb83's addition of the generalized inverse Gaussian distribution (https://github.com/scipy/scipy/pull/8681) ready for 1.3, so I might not be able to get to this for a while.

**Comment by pvanmulbregt** [comment]
@WarrenWeckesser thanks for the consideration and notification.

**Comment by pvanmulbregt** [comment]
Rebased again to get around a conflicting change in `scipy/stats/_distn_infrastructure.py`

**Comment by pvanmulbregt** [comment]
Rebased again to get around a conflicting change in `scipy/stats/_continuous_distns.py`

**Comment by chrisb83** [comment]
@pvanmulbregt You need to rebase since tests of the fit method have been moved in #10660

File: doc/source/tutorial/stats/continuous_kstwo.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
+
+.. _continuous-kstwo:
+
+KStwo Distribution
+==================
+
+This is the distribution of the maximum absolute differences between an
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_kstwo.rst]: Let's add that the underlying distribution of the samples is assumed to be continuous

File: doc/source/tutorial/stats/continuous_kstwo.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+>>> kstwo.sf([0, 0.5, 1.0], 5)
+array([1.   , 0.112, 0.   ])
+
+Compare a sample of size 5 drawn from a source N(0.5, 1) distribution against
+a target N(0, 1) CDF.
+
+>>> from scipy.stats import norm
+>>> n = 5
+>>> gendist = norm(0.5, 1)       # Normal distribution, mean 0.5, stddev 1
+>>> np.random.seed(seed=233423)  # Set the seed for reproducibility
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_kstwo.rst]: if we use this seed, the chart does not look very nice since Dn+ is too small to be shown.

By trying a few more seeds, I noted that 423 looks better :)

File: doc/source/tutorial/stats/continuous_kstwo.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ Kolmogorov-Smirnov 2-sided n=5: Prob(Dn >= 0.563278) = 0.0500
+
+Plot the Empirical CDF against the target N(0, 1) CDF
+
+>>> import matplotlib.pyplot as plt
+>>> plt.step(np.concatenate([[-3], x]), ecdfs, where='post', label='Empirical CDF')
+>>> x3 = np.linspace(-3, 3, 100)
+>>> plt.plot(x3, target.cdf(x3), label='CDF for N(0, 1)')
+>>> plt.ylim([0, 1]); plt.grid(True); plt.legend();
+# Add vertical lines marking Dn+ and Dn-
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_kstwo.rst]: suggest to remove # and add new lines before and after the text

Or add Dn+ and DN- to the legend?

File: doc/source/tutorial/stats/continuous_kstwo.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+>>> iminus, iplus = np.argmax(gaps, axis=0)
+>>> plt.vlines([x[iminus]], ecdfs[iminus], cdfs[iminus], color='r', linestyle='dashed', lw=4)
+>>> plt.vlines([x[iplus]], cdfs[iplus], ecdfs[iplus+1], color='m', linestyle='dashed', lw=4)
```
Comment (chrisb83) [inline doc/source/tutorial/stats/continuous_kstwo.rst]: minor Point: I found the colors hard to distinguish, maybe use linestyle='dotted' for one of the lines

File: scipy/stats/_continuous_distns.py [hunk 5]
```diff
@@ -113,6 +115,66 @@ def _isf(self, q, n):
 ksone = ksone_gen(a=0.0, b=1.0, name='ksone')
 
 
+class kstwo_gen(rv_continuous):
+    r"""General Kolmogorov-Smirnov two-sided test.
```
Comment (chrisb83) [inline scipy/stats/_continuous_distns.py]: this might be misleading since it is not a test, same in __init__.py

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -113,6 +115,66 @@ def _isf(self, q, n):
...
+
+    Notes
+    -----
+    :math:`D_n` is given by
+
+    .. math::
+
+        D_n &= \text{sup}_x |F_n(x) - F(x)|
+
+    where :math:`F` is a CDF and :math:`F_n` is an empirical CDF. `kstwo`
```
Comment (chrisb83) [inline scipy/stats/_continuous_distns.py]: continuous CDF

**Review by chrisb83 — COMMENTED** [review]
I started to review the documentation and added a few comments

File: doc/source/tutorial/stats/continuous_kstwo.rst (hunk 1 above)
Comment (pvanmulbregt) [reply doc/source/tutorial/stats/continuous_kstwo.rst]: Agreed.

File: doc/source/tutorial/stats/continuous_kstwo.rst (hunk 2 above)
Comment (pvanmulbregt) [reply doc/source/tutorial/stats/continuous_kstwo.rst]: Changed to 423.

File: doc/source/tutorial/stats/continuous_kstwo.rst (hunk 4 above)
Comment (pvanmulbregt) [reply doc/source/tutorial/stats/continuous_kstwo.rst]: 'dotted' didn't help but I added annotations to the plot which should make it clearer.

File: scipy/stats/_continuous_distns.py (hunk 5 above)
Comment (pvanmulbregt) [reply scipy/stats/_continuous_distns.py]: Agreed. Changed.

File: scipy/stats/_continuous_distns.py (hunk 6 above)
Comment (pvanmulbregt) [reply scipy/stats/_continuous_distns.py]: Agreed. Changed.

File: doc/source/tutorial/stats/continuous_kstwo.rst (hunk 3 above)
Comment (pvanmulbregt) [reply doc/source/tutorial/stats/continuous_kstwo.rst]: Removed the # and added some annotations to the plot.

File: scipy/stats/tests/test_distributions.py [hunk 7]
```diff
@@ -971,6 +971,109 @@ def test_stats(self):
         assert_allclose(result, expected)
 
 
+class TestKSTwo(object):
+    def setup_method(self):
+        np.random.seed(1234)
+
+    def test_cdf(self):
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: suggest to add a comment: test values for x close to 0 and 1 (exact values given by eq, 1 in Simard / L'Ecuyer);
for x = 0.5: I guess it is related to this identity
  Prob(D_n >= d) = 2*Prob(D_n+ >= d) - Prob(D_n+ >= d and D_n- >= d).
  For d > 0.5, the latter intersection probability is 0.

File: scipy/stats/tests/test_distributions.py [hunk 8]
```diff
@@ -971,6 +971,109 @@ def test_stats(self):
...
+        for n in [1, 2, 3, 10, 100, 1000]:
+            x = np.array([0, 0.5/n, 1/n, 0.5, 1-1.0/n, 1])
+            v1 = (1.0/n)**n
+            lg = scipy.special.gammaln(n+1)
+            elg = (np.exp(lg) if v1 != 0 else 0)
+            expected = np.array([1.0, 1.0,
+                                 1 - v1 * elg,
+                                 2*stats.ksone.sf(0.5, n),
+                                 min(2*v1, 1.0), 0])
+            vals_cdf = stats.kstwo.sf(x, n)
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: vals_sf, also next line

File: scipy/stats/tests/test_distributions.py [hunk 9]
```diff
@@ -971,6 +971,109 @@ def test_stats(self):
...
+            assert_allclose(vals_cdf, expected)
+
+    def test_cdf_sqrtn(self):
+        x = np.linspace(0, 2, 11)[1:]
+        # ns = [1, 2, 3, 10, 100, 1000]
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: remove this line?

File: scipy/stats/tests/test_distributions.py [hunk 10]
```diff
@@ -971,6 +971,109 @@ def test_stats(self):
...
+        ns = [50, 100, 200, 400, 1000, 2000]
+        for _x in x:
+            xn = _x / np.sqrt(ns)
+            probs = stats.kstwo.cdf(xn, ns)
+            diffs = np.diff(probs)
+            cond = (diffs >= 1e-8)
+            assert len(diffs[cond]) == 0
```
Comment (chrisb83) [inline scipy/stats/tests/test_distributions.py]: where does this result come from?

**Review by chrisb83 — COMMENTED** [review]
I started to look into the test cases.

Simard and L'Ecuyer list values they use for benchmarking in Table 1 of their paper. I think it would make sense to use (some of) these values in a separate test case. What do you think?

File: scipy/stats/tests/test_distributions.py (hunk 7 above)
Comment (pvanmulbregt) [reply scipy/stats/tests/test_distributions.py]: Yes.  Added some doc.

File: scipy/stats/tests/test_distributions.py (hunk 8 above)
Comment (pvanmulbregt) [reply scipy/stats/tests/test_distributions.py]: Yes.  Renamed.

File: scipy/stats/tests/test_distributions.py (hunk 9 above)
Comment (pvanmulbregt) [reply scipy/stats/tests/test_distributions.py]: Removed.

File: scipy/stats/tests/test_distributions.py (hunk 10 above)
Comment (pvanmulbregt) [reply scipy/stats/tests/test_distributions.py]: For fixed `a`, as `n->infinity`, `kstwo.cdf(a/sqrt(n), n) -> kstwo.cdf(a, n)`, and it does so monotonically increasing.  This check just ensures that the sequence is increasing, up to some small floating point jitter.  I've aded some comments.

**Comment by chrisb83** [comment]
Simard and L'Ecuyer list values they use for benchmarking in Table 1 and Tables 6-12 of their paper. I think it would make sense to use (some of) these values in separate test cases. What do you think?

Maybe we can try to get this PR into release 1.4.

**Comment by pvanmulbregt** [comment]
> Simard and L'Ecuyer list values they use for benchmarking in Table 1 and Tables 6-12 of their paper. I think it would make sense to use (some of) these values in separate test cases. What do you think?

It's a good idea.  I've found some differences between a few of the numbers in the Simard/L'Ecuyer tables and ones produced by the implementation to be a little bigger than expected and in the midst of understanding why.

**Comment by pvanmulbregt** [comment]
There  are some unexpected errors in the CI checks:

* Travis-CI fails building the refguide using Py3.6 with 4 `scipy.sparse` errors involving int64 and longlong such as 
```
Expected:    
  <2x3 sparse matrix of type '<class 'numpy.int64'>'      
   with 2 stored elements in Compressed Sparse Column format>
Got:    
  <2x3 sparse matrix of type '<class 'numpy.longlong'>'    	
   with 2 stored elements in Compressed Sparse Column format>
```

* Linux_Python_36_32bit_full has 11 failures in `scipy.optimize` (`scipy.optimize.tests.test_linprog.TestLinprogIPSparse` and `scipy.optimize.tests.test_trustregion_krylov`) and `scipy.special` (comparing with `mpmath`).

Are there any known issues with these tests?

**Comment by pvanmulbregt** [comment]
> > Simard and L'Ecuyer list values they use for benchmarking in Table 1 and Tables 6-12 of their paper. I think it would make sense to use (some of) these values in separate test cases. What do you think?
> 
> It's a good idea. I've found some differences between a few of the numbers in the Simard/L'Ecuyer tables and ones produced by the implementation to be a little bigger than expected and in the midst of understanding why.

@chrisb83 After creating an independent implementation of Durbin's algorithm in SageMath and using 300bit arithmetic I believe that a few of the entries in Table 1 are not always within rtol=1e-5.  I added a test with expected results based on the SageMath calculations.

**Comment by pvanmulbregt** [comment]
Rebased past fixes for Issue gh-11270.

**Comment by chrisb83** [comment]
@pvanmulbregt is the PR ready for a second review?

**Comment by pvanmulbregt** [comment]
> @pvanmulbregt is the PR ready for a second review?

@chrisb83 Yes.

File: scipy/stats/_distn_infrastructure.py [hunk 11]
```diff
@@ -1668,7 +1666,10 @@ def _ppf_single(self, q, *args):
                 right = left
                 left *= factor
             # left is now such that cdf(left) < q
-        if not right:  # i.e. self.b = inf
+
+        if _b < np.inf:
```
Comment (chrisb83) [inline scipy/stats/_distn_infrastructure.py]: what is the reason for this change?

**Comment by chrisb83** [comment]
All my previous comments have been addressed. The test cases are comprehensive and the fact that the values in Table 1 of Simard / L'Ecuyer can be reproduced is a very good sign. Let me go one more time through the main code.

File: scipy/stats/_distn_infrastructure.py (hunk 11 above)
Comment (pvanmulbregt) [reply scipy/stats/_distn_infrastructure.py]: Cleaned up `_ppf_single` to make its intent clearer.

**Review by chrisb83 — COMMENTED** [review]
I mainly reviewed `_kolmogn` for now and I have some questions / suggestions.

File: scipy/stats/_distn_infrastructure.py [hunk 12]
```diff
@@ -1654,25 +1654,20 @@ def _ppf_to_solve(self, x, q, *args):
...
+        if np.isinf(left):
+            left = min(-factor, right)
             while self._ppf_to_solve(left, q, *args) > 0.:
-                right = left
-                left *= factor
+                left, right = left * factor, left
             # left is now such that cdf(left) < q
-        if not right:  # i.e. self.b = inf
-            right = factor
+            # if right has changed, then cdf(rught) >= q
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_distn_infrastructure.py]: Fixed

File: scipy/stats/_ksstats.py [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+            expnt += Hexpnt
+        H = np.matmul(H, H)
+        Hexpnt *= 2
+        # Scale as needed.
+        if np.abs(H[k - 1, k - 1]) > _EP128:
+            H /= _EP128
+            Hexpnt += _E128
+        nn = nn // 2
+
+    # ans = (Hpwr[k-1, k-1] * 2^expnt) * n!/n^n
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Removed.

File: scipy/stats/_ksstats.py [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+    else:
+        # i + 1 = 2*id2 + remainder
+        id2 = (i + 1) // 2
+        if (i + 1) % 2 == 0:  # i+1 is even, so i is odd
+            if id2 == n + 1:
+                j1, j2 = n - ll - ceilf - 1, n + ll + ceilf - 1
+            else:
+                j1, j2 = id2 - 1 - ll - roundf - 1, id2 + ll - 1 + ceilf - 1
+        else:
+            assert (i + 1) % 2 != 0  # i+1 is odd, i is even
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Removed.

File: scipy/stats/_ksstats.py [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+        if nxsquared >= 18.0:
+            return 1.0
+        if n <= 100000:
+            if n * x**1.5 <= 1.4:
+                prob = _kolmogn_DMTW(n, x, cdf=True)
+                return _pin_prob(prob)
+            prob = _kolmogn_PelzGood(n, x, cdf=True)
+            return _pin_prob(prob)
+        # n > 1e5
+        if nxsquared >= 18.0:
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Removed.

File: scipy/stats/_ksstats.py [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+
+    Simard & L'Ecuyer (2011) [7]."""
+    if int(n) != n or n <= 0:
+        return np.nan
+    if x >= 1.0:
+        return _select_and_pin_prob(1.0, 0.0, cdf=cdf)
+    if x <= 0.0:
+        return _select_and_pin_prob(0.0, 1.0, cdf=cdf)
+    t = n * x
+    if t <= 1.0:  # Ruben-Gambino
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Ruben-Gambino certainly is valid for n>140. I replaced the product for n>=140 with a calculation using logs.

File: scipy/stats/_ksstats.py [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+    """Compute the CDF(SF) for the two-sided Kolmogorov-Smirnov statistic.
+
+    x must be of type float, n of type integer.
+
+    Simard & L'Ecuyer (2011) [7]."""
+    if int(n) != n or n <= 0:
+        return np.nan
+    if x >= 1.0:
+        return _select_and_pin_prob(1.0, 0.0, cdf=cdf)
+    if x <= 0.0:
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: x<= 1/(2n) is not quite the same as n*x<= 1/2, so I left it the way it is, as the later calculation needs (2*n*x-1) to be positive, so can't drop the test n*x<=0.5.

File: scipy/stats/_ksstats.py [hunk 18]
```diff
@@ -0,0 +1,607 @@
...
+        onemx = 1.0 - x
+        prob = 2 * onemx**n
+        return _select_and_pin_prob(1 - prob, prob, cdf=cdf)
+    if x >= 0.5:  # Exact: 2 * smirnov
+        prob = 2 * scipy.special.smirnov(n, x)
+        return _select_and_pin_prob(1.0 - prob, prob, cdf=cdf)
+
+    nxsquared = t * x
+    if n <= 140:
+        if nxsquared <= 0.754693:
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: These limits (0.754693, 4.0) came from section 4(i). Can you point me to the section where you see nxsquared <= 0.25 and n <= 140? I think that Ruben-Gambino has already been applied where it can be by the time the code gets here.

File: scipy/stats/_ksstats.py [hunk 19]
```diff
@@ -0,0 +1,607 @@
...
+            prob = _kolmogn_DMTW(n, x, cdf=True)
+            return _select_and_pin_prob(prob, 1.0 - prob, cdf=cdf)
+        if nxsquared <= 4:
+            prob = _kolmogn_Pomeranz(n, x, cdf=True)
+            return _select_and_pin_prob(prob, 1.0 - prob, cdf=cdf)
+        # Now use Miller approximation of 2*smirnov
+        prob = 2 * scipy.special.smirnov(n, x)
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: I used `special.smirnov` because it is available in SciPy and is at least as accurate as `1`.

File: scipy/stats/_ksstats.py [hunk 20]
```diff
@@ -0,0 +1,607 @@
...
+        return _select_and_pin_prob(1.0 - prob, prob, cdf=cdf)
+
+    # Split CDF and SF as they have different cutoffs on nxsquared.
+    if cdf:
+        if nxsquared >= 18.0:
+            return 1.0
+        if n <= 100000:
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Done

File: scipy/stats/_ksstats.py [hunk 21]
```diff
@@ -0,0 +1,607 @@
...
+    x must be of type float, n of type integer.
+
+    Simard & L'Ecuyer (2011) [7]."""
+    if int(n) != n or n <= 0:
+        return np.nan
+    if x >= 1.0:
+        return _select_and_pin_prob(1.0, 0.0, cdf=cdf)
+    if x <= 0.0:
+        return _select_and_pin_prob(0.0, 1.0, cdf=cdf)
+    t = n * x
```
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: Ack.

**Comment by pvanmulbregt** [comment]
The two Travis-ci test failures are due to gh-11460 and unrelated to this PR.

**Comment by pvanmulbregt** [comment]
Rebased after CI failures from gh-11460 were addressed.

File: scipy/stats/_ksstats.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: > Ruben-Gambino certainly is valid for n>140

Indeed, I just noted that Simard/L'Ecuyer use RG for n <= 140 only. I think it is a good idea to use the asymptotic expansion for n > 140. is this covered by one of the test cases?

File: scipy/stats/_ksstats.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: ok

File: scipy/stats/_ksstats.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: ok, there are more comments now that describe the different regions

File: scipy/stats/_ksstats.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: ok

File: scipy/stats/_ksstats.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: You are righr, RG has already been applied.

**Comment by chrisb83** [comment]
Hi Paul,

I am fine with the changes and I think all of my comments have been taken into account. The only open point is to clarify whether the approximation of the Ruben-Gambino formula is covered by a test case. I think that would be good since the implementation differs slighlty from [1].

File: scipy/stats/_ksstats.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (pvanmulbregt) [reply scipy/stats/_ksstats.py]: The detailed descriptions in Simard/L'Ecuyer don't always explicitly list usage of R-G. 
From Figures 2 and 3 it is not immediately obvious whether the R-G regions stop at n=140, or extend beyond. A very close inspection suggests that they do indeed extend, but the line is very close to the y-axis.

But in Section 2.1 they imply that they use R-G whenever they can:

> In the tails (for x close to 0 or 1), the exact values for arbitrary n are (Ruben and Gambino 1982):
> ... (Formulae)
> This also provides an exact formula for the complementary cdf Fn(x) = 1 − Fn(x) over the same range of values of (n, x). We will use these simple formulas whenever they apply.

I.e. Simard/L'Ecuyer use R-G whenever it can be applied, and this implementation does the same.

**Comment by pvanmulbregt** [comment]
See discussion above https://github.com/scipy/scipy/pull/9719#discussion_r383038304 as to whether this implementation and Simard-L;Ecuyer actually differ on use of Ruben-Gambino.

Independent of that discussion, `TestKSTwo.test_cdf` and `TestKSTwo.test_sf` both test the Ruben-Gambino case, for `n` above and below 140.

File: scipy/stats/_ksstats.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chrisb83) [reply scipy/stats/_ksstats.py]: OK, Section 2.1 is pretty clear whereas 4.ii and 4.iii does not mention R-G. Thinking about it again, using R-G makes sense since one can accuratly estimate the very small probabilities using the approximation you implemented. Sorry about the confusion.

**Review by chrisb83 — APPROVED** [review]
All my comments have been addressed.

**Comment by mdhaber** [comment]
@pvanmulbregt Thanks for this! Would you add this to the [1.5.0 release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.5.0)?

**Comment by pvanmulbregt** [comment]
> @pvanmulbregt Thanks for this! Would you add it to the [1.5.0 release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.5.0)?

Done.
